soseki/utils/data_utils.py

```diff
@@ -28,6 +34,132 @@ class RetrieverDatasetExample:
     metadata: Optional[Any]
 
 
+class MultipleKeyDataset(Dataset):
+    dataset_dict: Dict[str, Dataset]
+    dataset_keys: List[str]
+    dataset_sizes: Dict[str, int]
+    total_dataset_size: int
+
+    def __init__(self, dataset_dict: Dict[str, Dataset]) -> None:
+        super(MultipleKeyDataset, self).__init__()
+        self.dataset_dict = dataset_dict
+        self.dataset_keys = list(dataset_dict.keys())
+        self.dataset_sizes = {key: len(dataset) for key, dataset in dataset_dict.items()}
+        self.total_dataset_size = sum(self.dataset_sizes.values())
+
+    def __len__(self) -> int:
+        return self.total_dataset_size
+
+    def __getitem__(self, key_idx_tuple: Tuple[str, int]) -> Any:
+        dataset_key, sample_idx = key_idx_tuple
+        return self.dataset_dict[dataset_key][sample_idx]
+
+
+class MultipleKeyDatasetBatchSampler(Sampler[List[int]]):
+    def __init__(
+        self,
+        multiple_key_dataset: MultipleKeyDataset,
+        weights: Optional[Dict[str, float]] = None,
+        batch_size: int = 1,
+        shuffle: bool = True,
+        seed: int = 0,
+        distributed: bool = False,
+        num_replicas: Optional[int] = None,
+        rank: Optional[int] = None,
+    ) -> None:
+        if distributed:
+            if num_replicas is None:
+                if not dist.is_available():
+                    raise RuntimeError("Requires distributed package to be available")
+
+                num_replicas = dist.get_world_size()
+
+            if rank is None:
+                if not dist.is_available():
+                    raise RuntimeError("Requires distributed package to be available")
+
+                rank = dist.get_rank()
+
+            if rank >= num_replicas or rank < 0:
+                raise ValueError(
+                    "Invalid rank {}, rank should be in the interval [0, {}]".format(rank, num_replicas - 1)
+                )
+        else:
+            num_replicas = 1
+            rank = 0
+
+        if weights is None:
+            weights = dict()
+
+        for key in multiple_key_dataset.dataset_dict.keys():
+            weights.setdefault(key, 1.0)
+
+        self.multiple_key_dataset = multiple_key_dataset
+        self.weights = weights
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.seed = seed
+        self.distributed = distributed
+        self.num_replicas = num_replicas
+        self.rank = rank
+
+        self.chunk_size = self.batch_size * self.num_replicas
+        self.num_chunks = (
+            sum(
```

Comment:
This may be simply written as `sum(int(len(dataset) * self.weights[key]) // self.chunk_size for key, dataset in self.multiple_key_dataset.dataset_dict.items())`.